### PR TITLE
Add adaptive lightcurve time binning method

### PR DESCRIPTION
**Pull request by gabemery**
The goal of this addition is to produce time intervals which automatically adapt to varying signal.

Status : the interval creation seems to work (tested in an edited lightcurve notebook)
             work event by event in multiple observations
             possibility to force the start of a new point added (useful in case of large separation time between observations)
             no test yet  -> **now done**
             management of observation real start and end currently not done
             need some reflexion on precise interval definition (start on the event or between the event and the previous one,...)
             validity of change to event_list.py unchecked

Any input is welcome

### CODE REVIEW

**Comment by cdeil** [comment]
@gabemery - Thanks! Having an adaptive time binning method is a nice addition!

For each new function / method / class, the most important thing for us is to have an automated test that exercises the new code, and shows that it performs as expected.
I usually try to take an as simple as possible test case and add asserts at the end showing that it does what it's supposed to do at the end, because tring to do many tests that exercise all the edge cases leads to a ton of and diffcult to read and maintain test code.
Can you please add a test?

**Comment by gabemery** [comment]
The test is done and working
There seem to be problems when comparing astropy time object as they are approximated. It might be problematic in other areas of gammapy. (Here it changed the event content of each intervals and rejected one observation from a light curve point covering the transition between 2 observations)

Should be ready to merge (I will check if there is no temporary code left then edit this message)

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: Add empty line here. Sphinx is super picky about docstring formatting, this will error out in the docs build.

You can do this locally to check (but it's slow):
```
python setup.py build_docs
open docs/_build/html/index.html
```

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: Add space before the `:` or Sphinx won't format the docstrig correctly. Here and for all other parameters.

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: Add empty line before the Returns section heading of Sphinx won't format correctly.

**Review by cdeil — COMMENTED** [review]
Please check that your docstring is formatted correctly.

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: Suggest to just define this simple key function inline like this (and remove the `def gettime` above):
```
time_holder = sorted(time_holder, key=lambda item: item[0])
```

File: gammapy/time/tests/test_lightcurve.py
Comment (cdeil) [inline gammapy/time/tests/test_lightcurve.py]: Can you please add more direct asserts for the method / algorithm you're adding?
Assert on a few numbers from `intervals`.

Note that what you're writing here is a prototype, and then once we have figure out how we want to do Time object handling in Gammapy, and when refactoring your one super-long method into smaller functions, having a good test that establishes the desired output for one application is very valuable; even if for this test case the reference numbers that come out are arbitrary floats for now.

**Review by cdeil — COMMENTED** [review]
Would it make sense to use a list of dict where the dict contains a bit more infos, such as the n_on, n_off, alpha, significance, in addition to t_start and t_stop?

Then the tests could directly assert on the return value of this method, without wasting time to actually run the LC computation.

Note that if you store everything in one list of dicts, it's one line to make a Table and return a Table from your method if you prefer:
http://docs.gammapy.org/dev/notebooks/astropy_introduction.html#Creating-tables

@gabemery or @jlenain - Would you be available tomorrow or next week for a short Skype call with me to discuss this, before we merge it in?
If no, we can also finalise this via Github.

**Comment by gabemery** [comment]
From a "basic user" point of view, I think that the use of a list of dict would not make sense as a default output. The focus of the method is to provide a list of intervals to compute a light curve. A list of dict would force the user to re-extract the information.

Having access to the information is still interesting (mostly for test and debugging purpose) so I think that I could add a variable which change the output like so : 
- if false(default) : return intervals
- if true : return listofdict( tstart,tstop,alpha, non,noff,significance)  (or return intervals,listofdict)
This could be in another method recovering the listofdict.

The question remain on the efficiency of such an addition. I will do some testing.

Having tests comparing the lightcurve and the method output seems important as it can highlight problems from one or the other. Direct tests on the output can be added.

**Comment by cdeil** [comment]
> From a "basic user" point of view, I think that the use of a list of dict would not make sense as a default output.

@gabemery - Agreed. But I do think returning the info in a `Table` would be better than the list of time intervals we've been using so far.

We haven't discussed this yet, but my thinking so far was to use time interval tables (probably using the [GTI](http://docs.gammapy.org/dev/api/gammapy.data.GTI.html) class) and to build a set of GTI tools to split / merge / compute time intervals that are then used by AGN and binary people. I think that's also how current tools (e.g. X-ray analysis) work, users get data that includes GTI tables, and then they manipulate the GTIs to cover their science analysis cases.

**Comment by gabemery** [comment]
So if I use the list of dictionaries the execution time seem not to change. Here is what I tried:

- create_fixed_significance_bin_lc_dict  : return a list of dictionaries containing n_on, n_off, alpha, significance, in addition to t_start and t_stop

- extract_intervals : take the list and create a list of intervals like currently used for the lc

- create_fixed_significance_bin_lc : call both functions (see after)


def create_fixed_significance_bin_lc(self, significance, significance_method, energy_range, spectrum_extraction, `separators=[]):

return LightCurveEstimator.extract_intervals(self.create_fixed_significance_bin_lc_dict(significance,significance_method,energy_range, spectrum_extraction, separators))


I can push this if you prefer

**Comment by cdeil** [comment]
My suggestion would be that you put your list of dict with all the infos into a Table, as shown here:
http://docs.gammapy.org/dev/notebooks/astropy_introduction.html#Creating-tables
and return that from your function. This doesn't improve or decrease performance, that's just an API change / preference.

Then in your test, remove the call to  `c_estimator.light_curve` and only run the new method of interest, and assert on some values you get back in the table. This will make your test run a little bit faster. You can see the time of your test by adding the `--durations=0` option.

**Review by cdeil — CHANGES_REQUESTED** [review]
@gabemery and I talked about this a bit on Skype. I left some inline comments.

Let me know if you have any questions or when I should review again.

File: gammapy/data/event_list.py
Comment (cdeil) [inline gammapy/data/event_list.py]: I'd suggest to rename to `observation_time_start` and `observation_time_end` for now, for consistency with the other observation-related properties: http://docs.gammapy.org/dev/api/gammapy.data.EventList.html

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: The name and API of the methods to make time intervals should be consistent.
How about this?

* `create_fixed_time_bin` -> `make_time_intervals_fixes`
* `create_fixed_significance_bin_lc_table` -> `make_time_intervals_min_significance`

And change the existing method to return a table with start and stop time in the same format that we use here.
I.e. after one has computed the time interval, the rest of the code would be the same for users.

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: Change `separators=[]` to `separators=None` and put the default in the method.
Read this: https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: Remove empty line above this one, and remove the "and store information".

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: From copy & pasting from the HTML docs you have non-ascii quote characters here.
Put normal single quotes:
```
        significance_method : {'lima', 'simple'}
```

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: There's a missing `_` here and the examples section isn't properly formatted. Put this:
```
        Examples
        --------
        Sometimes it can be useful to have the time intervals as a list of tuples
        of start and stop time. You can get that from the table like this::

            intervals = list(zip(table['t_start'], table['t_stop']))
```

File: gammapy/time/tests/test_lightcurve.py
Comment (cdeil) [inline gammapy/time/tests/test_lightcurve.py]: This line is too long; suggest to break it like this:
```
    table = lc_estimator.create_fixed_significance_bin_lc_table(
        significance=3, significance_method='lima',
        energy_range=[0.2, 100] * u.TeV,
        spectrum_extraction=spec_extract,
    )
```

**Review by cdeil — CHANGES_REQUESTED** [review]
@gabemery - I tried your branch locally and have some suggestions (see inline comments).

One small problem is that your test is pretty slow; this shows that it runs for six seconds:
```
python setup.py build_ext -i
python -m pytest -v gammapy/time/tests/test_lightcurve.py --duration=0
```
Another more important issue is that your current test case doesn't pass `separators` and thus the functionality related to `separators` is completely untested. I saw this by running
```
python setup.py test -V -t gammapy/time/tests/test_lightcurve.py  --coverage
open htmlcov/index.html
```
and then the lines of the algorithm related to separator handling show up in red as never executed.

The ideal solution here would be to write a function "make_example_dataset" that produces a great test case for all the LC functionality (e.g. have ~ 3 runs with each ~ 10 events), designed to test all the functionality / cases and be super fast to execute. But that's ~ 50 lines to write, and not very easy. So for this PR, can you stick with the current example, but improve it a bit e.g. introduce an arbitrary separator to show that that code is working and cover it with a test?

File: gammapy/time/lightcurve.py
Comment (jlenain) [inline gammapy/time/lightcurve.py]: For better readability later, it might be useful to quickly document the structure of `time_holder`, e.g. stating which "type" (`start`, `end`, `break`, `on`, `off`) the second argument could be.

File: gammapy/time/lightcurve.py
Comment (jlenain) [inline gammapy/time/lightcurve.py]: Is it necessary to call `select_energy` twice ?

**Review by jlenain — COMMENTED** [review]
Hi @gabemery ,

I am starting the review just now, sorry for the delay.
@cdeil , you already provided quite extensive feedbacks to @gabemery and I do not have much to add at this stage.

File: gammapy/time/lightcurve.py
Comment (gabemery) [reply gammapy/time/lightcurve.py]: good idea, I will this in the docstring

File: gammapy/time/lightcurve.py
Comment (gabemery) [reply gammapy/time/lightcurve.py]: I don't think so seeing the definition of emin and emax. But since I'm not sure I keeped what is in place

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: The docstring is what users see. So ideally, here you would give a high-level description of the algorithm, but not describe implementation details e.g. that there's an internal variable `time_holder`. For such a complex algorithm it can be useful to have a few line of comments with `# ...` at the start of the function that describe the idea, i.e. that the central data structure you create and operate on is the `time_holder` list.

Some small suggestions:

- Make `separators` a `Time` object? And better describe what it's purpose is in the docstring for users. How would one usually use it? Does it force breaks even if the desired significance is not reached?

- Rename `time_holder` variable to `events`?

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: I think here you're executing `time.tt.mjd` for every single event?
This is pretty inefficient, especially if the `time` timescale is not `tt` and a time scale transformation has to be computed.
To avoid it you can do `time = time.tt.mjd` at the array level once at the start, no?

File: gammapy/time/lightcurve.py
Comment (cdeil) [inline gammapy/time/lightcurve.py]: I guess it's not a real efficiency concern here, but still, I'd suggest you just store the floats in the dict here, and then after you've made your table, convert back to Time objects with one call processing arrays.
```
table = Table(rows=rows)
table['t_start'] = Time(table['t_start'], format='mjd', scale='tt')
table['t_stop'] = Time(table['t_stop'], format='mjd', scale='tt')
```

File: gammapy/time/tests/test_lightcurve.py
Comment (cdeil) [inline gammapy/time/tests/test_lightcurve.py]: IMO you can see here that just using floats as `separator` leads to slightly less readable code in callers.
You have:
```
separator = [(53343.94050200008 + 53343.952979345195) / 2]
```
If you use `Time` objects and change the parameter name to include the word "time", it would look like this:
```
time_split_points = Time([(53343.94050200008 + 53343.952979345195) / 2], scale='tt', format='mjd')
```
or you could directly work with more human-readable times in the tests if you like.

**Review by cdeil — CHANGES_REQUESTED** [review]
This is coming along nicely!

I left some more inline comments.

One thing that has to be fixed before we merge is the CI fail, you have to adapt the method name in the test here:
https://travis-ci.org/gammapy/gammapy/jobs/373066559#L3889

Another point to consider before merging is the version history, because once a pull request is merged, the commits will forever be part of the Gammapy version history. As you can see at https://github.com/gammapy/gammapy/pull/1390/commits it's non-ideal in this case, because there was some experiementation and you have change to `time.value` in unrelated parts of the code and then back again, and now you've merge upstream master into this branch, which often leads to problems and you should have rebased instead. See http://docs.astropy.org/en/stable/development/workflow/development_workflow.html#astropy-guidelines-for-git

> Never merge changes from astropy/master into your feature branch. If changes in the development version require changes to our code you can Rebase, but only if asked.

So overall my preference would be that you close this PR and start a new one from latest upstream master, copying over the files you have here and committing them with one commit per file (first the `event_list.py`, then `lightcurve.py`, then `test_lightcurve.py`). Then we would get a clean and simple version history in Gammapy.

If you don't want to do that here, OK, I know it's a bit of a hassle and complicated if you're new to git & Github. But then for future PRs, please avoid `git merge` upstream master into feature branches, and instead follow the Astropy workflow advice to `git rebase`, but only if needed.

File: gammapy/time/lightcurve.py
Comment (gabemery) [reply gammapy/time/lightcurve.py]: I can make it a TIme object. Changed the docstring. It seem there is a confusion, a bin can't reach the wanted significance and a break, so the desired significance is never reached. The current point is discarded when reaching a 'break' element.

Not all element of time_holder is an event so I'm not for this change.

File: gammapy/time/lightcurve.py
Comment (gabemery) [reply gammapy/time/lightcurve.py]: I tried at array level but it didn't work.
Isn't .tt.mjd equivalent to .value but with the added correction in case of different scales?

File: gammapy/time/lightcurve.py
Comment (gabemery) [reply gammapy/time/lightcurve.py]: Is it useful? Because if it change nothing it would only add 3 lines of code

File: gammapy/time/lightcurve.py
Comment (cdeil) [reply gammapy/time/lightcurve.py]: I think `.value` can give floats, or strings, or datetime objects, depending on the `format` of the time object.
```
In [3]: Time.now()
Out[3]: <Time object: scale='utc' format='datetime' value=2018-05-02 10:39:04.151454>

In [4]: Time.now().value
Out[4]: datetime.datetime(2018, 5, 2, 10, 39, 10, 801319)
```
So I don't think just calling `time.value` is a good idea.

I think either `time.tt.mjd` for all times (from the events and lightcurve, and user inputs) is OK, or probably an even better general strategy for Gammapy would be to go to "MET", i.e. mission elapsed time in seconds when working with times as floats, i.e. do `met = (time - events.time_ref).sec` and work with those floats. This is what I would do when working with event times; it's just that for lightcurves I think people often use MJD, i.e. days with a different fixed reference time that doesn't depend on the event list. So I'm not sure here.

But anyways, my comment was to avoid this call for each event separately, but do it for the array of events at the top. That should work just fine (you just say "didn't work") either way and be more effiicent and simpler that way, because it separates the time object (scale and format) handling from the algorithm, but doing a sandwich (Time -> float at start, then the algorithm, then float -> Time at the end). It's not super important, but I do think it would be overall cleaner to write it that way.

File: gammapy/time/lightcurve.py
Comment (cdeil) [reply gammapy/time/lightcurve.py]: See my comment above. I do think structuring the code to do Time <-> float only at the start and end would be a bit simpler, and faster than having conversions in the middle of the algorithm. But I agree it's not a huge win, more a small preference. So you decide.

File: gammapy/time/lightcurve.py
Comment (gabemery) [reply gammapy/time/lightcurve.py]: I will try again, but if I can't make it work  I will keep it as it is.

**Pull request closed without merging**
